core/trino-main/src/main/java/io/trino/connector/DefaultCatalogFactory.java

```
@@ -144,6 +146,25 @@ public CatalogConnector createCatalog(CatalogHandle catalogHandle, ConnectorName
         return createCatalog(catalogHandle, connectorName, connector, Optional.empty());
     }
 
+    @Override
+    public Set<String> getSecuritySensitivePropertyNames(CatalogProperties catalogProperties)
+    {
+        ConnectorFactory connectorFactory = connectorFactories.get(catalogProperties.connectorName());
+        if (connectorFactory == null) {
+            // If someone tries to use a non-existent connector, we assume they
```

This should be an error, actually. It should throw IllegalArgumentException.

---

But then the query will fail during redaction. The idea is to avoid disrupting the natural flow and let it fail where it normally would if redaction didn't exist.

---

Why would a query fail during redaction if it hasn’t first failed during analysis? I.e., it’s a condition that should never occur.

---

We perform redaction at a very early stage (before the query state machine is created) to modify query text exposed in query events and `QueryInfo`. I believe that verifying the existence of a given connector happens only during execution, for example, in `CreateCatalogTask`.

---

I’m very confused about the purpose of this change, then. It redaction happens before analysis, how is the analyzer and execution engine able to see the unredacted values so that it can to its job?

Can you describe the technical approach at a high level so that I don’t have to reverse engineer what the code is trying to achieve?

---

Please let me know if the following helps:

### Problem
Currently, when we execute a CREATE CATALOG statement containing plaintext secrets, unredacted query text is exposed via the REST API, the `system.runtime.queries table`, and query events.

### Goal
Instead of displaying the query text in its raw form in the locations mentioned above, such as:

```sql
CREATE CATALOG test USING postgresql  
WITH (  
   "connection-user" = 'bob',  
   "connection-password" = '1234'  
)  
```

we aim to redact security-sensitive property values:

```sql
CREATE CATALOG test USING postgresql  
WITH (  
   "connection-user" = 'bob',  
   "connection-password" = '***'  
)  
```

### Proposed Solution
The REST API, the `system.runtime.queries` table, and query events obtain query text from the `QueryInfo` object. Based on our research, the query text contained in `QueryInfo` is not interpreted anywhere in the engine.

The `QueryInfo` object is created by the `QueryStateMachine`. To redact the query text, we propose performing redaction after the query is parsed (to ensure we have the AST, available for traversal and redaction) but before the `QueryStateMachine` is created.

Since redaction occurs at an early stage of query processing, we need to duplicate some logic that is typically performed during analysis and execution. For example, this includes evaluating catalog properties. Additionally, we do not want to disrupt the normal query processing flow; therefore, we ensure the query never fails due to redaction. If, for any reason, redaction is not possible, we will resort to masking all properties.

To identify security-sensitive properties for a given connector, we propose introducing a new SPI to expose them: https://github.com/trinodb/trino/pull/24562